openwillis/measures/video/preprocess_video.py

```diff
@@ -0,0 +1,815 @@
+            facedata_df.cluster==cluster_idx
+        ]
+        
+        # clean face data
+        cluster_df = filter_cluster_segments(
+            cluster_df,
+            min_frames_face_present,
+            frames_per_row,
+            fps
+        )
+
+        # upsample to frames and convert to dict list
+        face_bbox_list = convert_face_df_to_bbox_list(
+            cluster_df,
+            frames_per_row,
+            num_frames_vid,
+            max_frame_idx
+        )
+
+        face_list_dict[cluster_idx] = face_bbox_list
+
+    return face_list_dict
+
+def preprocess_face_video(
+    video_path,  
+    capture_n_frames_per_second=2,
+    model_name='Facenet',
+    detector_backend='mtcnn', 
+    threshold=.95, 
+    min_sec_face_present=3, 
+    n_clusters=2,
+    n_frames=np.inf,
+    darken=True,
+    keep_original_timing=False,
+    default_size_for_cropped = (512,512)
+):
+    """
+    Preprocesses a face video by extracting face data, clustering the faces, and preparing the output.
+
+    Parameters:
+    - video_path (str): Path to the input video file.
+    - capture_n_frames_per_second (int): Number of frames to capture per second. Increasing this parameter should increase the quality of clustering but will cause the function to run for longer. Default is 2.
+    - model_name (str): Name of the face recognition model to use. Default is 'Facenet'.
+    - detector_backend (str): Backend to use for face detection. Can be any model used in the Deepface library. Default is 'mtcnn'.
+    - threshold (float): Similarity threshold for clustering faces. Default is 0.95.
+    - min_sec_face_present (int): Minimum number of seconds a face must be present after clustering to not be filtered out. Default is 3.
+    - n_clusters (int): Number of clusters to create. Default is 2, assuming this is a clinical interview.
+    - n_frames (int): Maximum number of frames to process. Default is np.inf (i.e. process all frames).
+    - darken (bool): Whether to darken the cropped video all pixels out of face bounding box - but leave face at same location in frame. If false only frames within bounding box plus standardized boarder are returned. Default is True.
+    - keep_original_timing (bool): Whether to keep the original timing for the cropped video. If true all frames are retained
+      but frames with out a face of the identity are returned as black, if false only frames with the identity are returned. Default is False.
+    - default_size_for_cropped (tuple): Default size for the cropped video. Default is (512, 512).
+
+    Returns:
+    - fps (float): Frames per second of the processed video. Used for saving video
+    - cropped_video_list (list): List of cropped videos.
+    - out_dict (dict): Dictionary containing the preprocessed face data.
+    - facedata_df (pandas.DataFrame): DataFrame containing the face data and cluster info.
+    """
+    config = get_config(os.path.abspath(__file__), 'preprocess.json')
+
```

Comment:
To ensure no error and that output is produced even if the main functionality fails, you should set your outputs values to something meaningful in the beginning. So even if the `try` statement fails something is outputted.

Especially because this is going to be part of a pipeline and post-processing is to be conducted, what should be returned in case of fail for each argument is worth a think.

Reply:
I'm thinking these as the defaults:

fps = -1
facedata_df = pd.DataFrame()
out_dict = {}
cropped_video_dict = {}

fps as -1 because it is sort of a classic error return and clearly indicates fps was never set

facedata_df as an empty pandas df because it indicates you never got there - the other option I considered for this is to have it with the same columns as the returned df but empty rows - I opted away from this because you would get that same return (i.e. columns with 0 rows) if something was messed up in the filtering or the detection algorithms didn't find any faces (above threshold).

out_dict as an empty dict for the same reasons above

cropped_video_dict as an empty dict for the same reasons above - note I also updated it from a list to a dict for consistency with out_dict.

If those all seem reasonable I will add as a commit.

Reply:
Yes seem reasonable to me